File: server/supabase/Geo.swift
import Foundation

/// Geographic helpers for the nearby-photo search.
///
/// Two-stage search: a cheap bounding-box prefilter PostgreSQL can answer from
/// an index, then an exact haversine pass in Swift. Same shape as server/geo.js.
enum Geo {
    private static let earthRadiusM = 6371008.8

    /// Great-circle distance between two coordinates, in metres.
    static func haversineM(lat1: Double, lon1: Double, lat2: Double, lon2: Double) -> Double {
        let dLat = (lat2 - lat1) * .pi / 180
        let dLon = (lon2 - lon1) * .pi / 180
        let a = sin(dLat / 2) * sin(dLat / 2)
            + cos(lat1 * .pi / 180) * cos(lat2 * .pi / 180) * sin(dLon / 2) * sin(dLon / 2)
        return 2 * earthRadiusM * asin(min(1, sqrt(a)))
    }

    /// Bounding box that fully contains the circle of `radiusM` around a point.
    static func bboxFor(lat: Double, lon: Double, radiusM: Double) -> BoundingBox {
        let latDelta = (radiusM / earthRadiusM) * (180 / .pi)
        let cosLat = cos(lat * .pi / 180)
        let lonDelta = abs(cosLat) < 1e-6
            ? 180.0
            : (radiusM / (earthRadiusM * cosLat)) * (180 / .pi)
        return BoundingBox(
            minLat: lat - latDelta,
            maxLat: lat + latDelta,
            minLon: lon - min(180, abs(lonDelta)),
            maxLon: lon + min(180, abs(lonDelta))
        )
    }

    /// Offset a coordinate by a distance and bearing. Used to scatter seed photos.
    static func offsetCoordinate(
        lat: Double,
        lon: Double,
        distanceM: Double,
        bearingDeg: Double
    ) -> (lat: Double, lon: Double) {
        let bearing = bearingDeg * .pi / 180
        let angular = distanceM / earthRadiusM
        let lat1 = lat * .pi / 180
        let lon1 = lon * .pi / 180
        let lat2 = asin(
            sin(lat1) * cos(angular) + cos(lat1) * sin(angular) * cos(bearing)
        )
        let lon2 = lon1 + atan2(
            sin(bearing) * sin(angular) * cos(lat1),
            cos(angular) - sin(lat1) * sin(lat2)
        )
        let normalizedLon = ((lon2 * 180 / .pi + 540).truncatingRemainder(dividingBy: 360)) - 180
        return (lat: lat2 * 180 / .pi, lon: normalizedLon)
    }
}

struct BoundingBox {
    let minLat: Double
    let maxLat: Double
    let minLon: Double
    let maxLon: Double
}

File: server/geo.js
// Geographic helpers for the nearby-photo search.
//
// The search is deliberately two-stage: a cheap bounding-box prefilter that
// SQLite can answer from an index, then an exact haversine pass in JS. At
// hackathon data volumes either stage alone would do, but this is the shape a
// real implementation wants, and it keeps the SQL portable (no spatial ext).

const EARTH_RADIUS_M = 6371008.8;

const toRad = (deg) => (deg * Math.PI) / 180;

/** Great-circle distance between two coordinates, in metres. */
export function haversineM(lat1, lon1, lat2, lon2) {
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) ** 2;
  return 2 * EARTH_RADIUS_M * Math.asin(Math.min(1, Math.sqrt(a)));
}

/**
 * Bounding box that fully contains the circle of `radiusM` around a point.
 * Longitude degrees shrink with latitude, so the lon delta is scaled by
 * 1/cos(lat); near the poles that blows up, so it is clamped to a full sweep.
 */
export function bboxFor(lat, lon, radiusM) {
  const latDelta = (radiusM / EARTH_RADIUS_M) * (180 / Math.PI);
  const cosLat = Math.cos(toRad(lat));
  const lonDelta =
    Math.abs(cosLat) < 1e-6
      ? 180
      : (radiusM / (EARTH_RADIUS_M * cosLat)) * (180 / Math.PI);
  return {
    minLat: lat - latDelta,
    maxLat: lat + latDelta,
    minLon: lon - Math.min(180, Math.abs(lonDelta)),
    maxLon: lon + Math.min(180, Math.abs(lonDelta)),
  };
}

/** Offset a coordinate by a distance and bearing. Used to scatter seed photos. */
export function offsetCoordinate(lat, lon, distanceM, bearingDeg) {
  const bearing = toRad(bearingDeg);
  const angular = distanceM / EARTH_RADIUS_M;
  const lat1 = toRad(lat);
  const lon1 = toRad(lon);
  const lat2 = Math.asin(
    Math.sin(lat1) * Math.cos(angular) +
      Math.cos(lat1) * Math.sin(angular) * Math.cos(bearing),
  );
  const lon2 =
    lon1 +
    Math.atan2(
      Math.sin(bearing) * Math.sin(angular) * Math.cos(lat1),
      Math.cos(angular) - Math.sin(lat1) * Math.sin(lat2),
    );
  return {
    lat: (lat2 * 180) / Math.PI,
    lon: (((lon2 * 180) / Math.PI + 540) % 360) - 180,
  };
}
